pkg/eth/transformer.go

```diff
@@ -0,0 +1,355 @@
+	"github.com/ethereum/go-ethereum/statediff"
+	node "github.com/ipfs/go-ipld-format"
+	"github.com/jmoiron/sqlx"
+	"github.com/multiformats/go-multihash"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/ipld-eth-indexer/pkg/ipfs/ipld"
+	"github.com/vulcanize/ipld-eth-indexer/pkg/postgres"
+	"github.com/vulcanize/ipld-eth-indexer/pkg/shared"
+)
+
+// Transformer interface to allow substitution of mocks for testing
+type Transformer interface {
+	Transform(workerID int, payload statediff.Payload) (int64, error)
+}
+
+// StateDiffTransformer satisfies the Transformer interface for ethereum statediff objects
+type StateDiffTransformer struct {
+	chainConfig *params.ChainConfig
+	indexer     *CIDIndexer
+}
+
+// NewStateDiffTransformer creates a pointer to a new PayloadConverter which satisfies the PayloadConverter interface
+func NewStateDiffTransformer(chainConfig *params.ChainConfig, db *postgres.DB) *StateDiffTransformer {
+	return &StateDiffTransformer{
+		chainConfig: chainConfig,
+		indexer:     NewCIDIndexer(db),
+	}
+}
+
+// Transform method is used to process statediff.Payload objects
+// It performs the necessary data conversions and database persistence
+func (sdt *StateDiffTransformer) Transform(workerID int, payload statediff.Payload) (int64, error) {
+	// Unpack block rlp to access fields
+	block := new(types.Block)
+	if err := rlp.DecodeBytes(payload.BlockRlp, block); err != nil {
+		return 0, err
+	}
+	blockHash := block.Hash()
+	logrus.Infof("worker %d transforming state diff payload for blocknumber %d with hash %s", workerID, block.Number().Int64(), blockHash.String())
+	transactions := block.Transactions()
+	// Block processing
+	// Decode receipts for this block
+	receipts := make(types.Receipts, 0)
+	if err := rlp.DecodeBytes(payload.ReceiptsRlp, &receipts); err != nil {
+		return 0, err
+	}
+	// Derive any missing fields
+	if err := receipts.DeriveFields(sdt.chainConfig, blockHash, block.NumberU64(), transactions); err != nil {
+		return 0, err
+	}
+	// Generate the block iplds
+	headerNode, uncleNodes, txNodes, txTrieNodes, rctNodes, rctTrieNodes, err := ipld.FromBlockAndReceipts(block, receipts)
+	if err != nil {
+		return 0, err
+	}
+	if len(txNodes) != len(txTrieNodes) && len(rctNodes) != len(rctTrieNodes) && len(txNodes) != len(rctNodes) {
+		return 0, fmt.Errorf("expected number of transactions (%d), transaction trie nodes (%d), receipts (%d), and receipt trie nodes (%d)to be equal", len(txNodes), len(txTrieNodes), len(rctNodes), len(rctTrieNodes))
+	}
+	// Begin new db tx for everything
+	tx, err := sdt.indexer.db.Beginx()
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		if p := recover(); p != nil {
+			shared.Rollback(tx)
+			panic(p)
+		} else if err != nil {
+			shared.Rollback(tx)
+		} else {
+			err = tx.Commit()
+		}
+	}()
+
+	// Publish and index header, collect headerID
+	reward := CalcEthBlockReward(block.Header(), block.Uncles(), block.Transactions(), receipts)
```

Comment:
we can start db transaction after this function `CalcEthBlockReward`. it doesn't write to the database but takes open transaction time. I guess it is probably some milliseconds but some sort of optimization

Reply:
Good catch, will do!